scripts/build/check_init_priorities.py

```diff
@@ -0,0 +1,316 @@
+
+    def __hash__(self):
+        return self._level_priority
+
+
+class ZephyrObjectFile:
+    """Load an object file and finds the device defined within it.
+
+    Load an object file and scans the relocation sections looking for the known
+    ones containing initialization callbacks. Then finds what device ordinals
+    are being initialized at which priority and stores the list internally.
+
+    Attributes:
+        file_path: path of the file to be loaded.
+    """
+    def __init__(self, file_path):
+        self.file_path = file_path
+        self._elf = ELFFile(open(file_path, "rb"))
+        self._load_symbols()
+        self._find_defined_devices()
+
+    def _load_symbols(self):
+        """Initialize the symbols table."""
+        self._symbols = {}
+
+        for section in self._elf.iter_sections():
+            if not isinstance(section, SymbolTableSection):
+                continue
+
+            for num, sym in enumerate(section.iter_symbols()):
+                if sym.name:
+                    self._symbols[num] = sym.name
+
+    def _device_ord_from_rel(self, rel):
+        """Find a device ordinal from a device symbol name."""
+        sym_id = rel["r_info_sym"]
+        sym_name = self._symbols.get(sym_id, None)
+
+        if not sym_name:
+            return None
+
+        if not sym_name.startswith(_DEVICE_ORD_PREFIX):
+            return None
+
+        _, device_ord = sym_name.split(_DEVICE_ORD_PREFIX)
+        return int(device_ord)
+
+    def _find_defined_devices(self):
+        """Find the device structures defined in the object file."""
+        self.defined_devices = {}
+
+        for section in self._elf.iter_sections():
+            if not isinstance(section, RelocationSection):
+                continue
+
+            if not section.name.startswith(_INIT_SECTION_PREFIX):
+                continue
+
+            prio = Priority(section.name)
+
+            for rel in section.iter_relocations():
+                device_ord = self._device_ord_from_rel(rel)
+                if not device_ord:
+                    continue
+
+                if device_ord in self.defined_devices:
+                    raise ValueError(f"Device {device_ord} already defined")
+
+                self.defined_devices[device_ord] = prio
+
+    def __repr__(self):
+        return (f"<ZephyrObjectFile {self.file_path} "
+                f"defined_devices: {self.defined_devices}")
+
+class Validator():
+    """Validates the initialization priorities.
+
+    Scans through a build folder for object files and list all the device
+    initialization priorities. Then compares that against the EDT derived
+    dependency list and log any found priority issue.
+
+    Attributes:
+        build_dir: the build directory to scan
+        log: a logging.Logger object
+    """
+    def __init__(self, build_dir, log):
+        self.log = log
+
+        edtser = os.path.join(build_dir, "zephyr", "edt.pickle")
```

Comment:
I think this path should come from the `EDT_FILE` variable in cmake instead of hard-coding the path?

Reply:
Good idea, added a flag.